Meraki - Convert response keys to snake_case from camelCase

kbreit opened the pull request:
##### SUMMARY
Current Meraki modules respond with camelCase and not Ansible's snake_case format. This proposal is one option of how to fix this.

- `output_version` parameter dictates the response key case
- `new is snake_case, `old` is camelCase
- If `new`, conversion is done at the end of module execution
- This is purely a proposal and not a final draft

I am hoping this can be integrated in Ansible 2.9. This would mean the parameter would be in operation until Ansible 2.12 for a 4 year change schedule. Currently this PR defaults to `new` so it would break existing playbooks which use the keys. However, it could easily be changed to `old` for the 2.9 release.

@felixfontein I'm mrproper from IRC. This is what we were discussing earlier. I thought of this idea and wanted to get your feedback on it. I am open to doing an `update()` on `data` to add both camelCase and snake_case keys, but that would be a lot of duplicate data and verbose. I'm not saying this is the best option either, but wanted to put the idea out there.

Fixes #53598

### Integration Tests Updated
- [x] meraki_network
- [x] meraki_organization
- [x] meraki_ssid
- [x] meraki_admin
- [x] meraki_config_template
- [x] meraki_device
- [x] meraki_switchport
- [x] meraki_mr_l3_firewall
- [x] meraki_mx_l3_firewall
- [x] meraki_content_filtering
- [x] meraki_snmp
- [x] meraki_syslog
- [x] meraki_vlan

### Documentation Updated
- [x] meraki_network
- [x] meraki_organization
- [x] meraki_ssid
- [x] meraki_admin
- [x] meraki_config_template
- [x] meraki_device
- [x] meraki_switchport
- [x] meraki_mr_l3_firewall
- [x] meraki_mx_l3_firewall
- [x] meraki_content_filtering
- [x] meraki_snmp
- [x] meraki_syslog
- [x] meraki_vlan


##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
meraki

kbreit committed:
Initial proposal for new parameter option for response format
- output_version parameter dictates the response key case
- new is snake_case, old is camelCase
- If new, conversion is done at the end of module execution
- This is purely a proposal and not a final draft

felixfontein commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -45,6 +46,7 @@ def meraki_argument_spec():
                 timeout=dict(type='int', default=30),
                 org_name=dict(type='str', aliases=['organization']),
                 org_id=dict(type='str'),
+                output_version=dict(type='str', choices=['old', 'new'], default='new'),
```
It's probably better to use more specific names than `old` and `new`, in case you want to add something else in the future. Using `new2` or `newer` will look really strange ;)

kbreit replied:
Old, new, newer, newest, newestAF...I dont see the big deal :)

Any ideas on replacements? I commented below with the possibility of snake and camel. But those aren’t obvious.

felixfontein replied:
That's a very good question. Maybe an integer? Version 1 is the old format, version 2 the new. Not a very helpful notation, though. And specifying a long string is also annoying.

kbreit replied:
I removed this option and moved it to an env var instead.

felixfontein commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -127,6 +130,26 @@ def define_protocol(self):
+                    items[snake_k] = self.sanitize_keys(data[k])
+            return items
+        elif isinstance(data, list):
+            items = []
```
Why not use a list comprehension here? It's probably more efficient:
`return [self.sanitize_keys(i) for i in data]`

kbreit replied:
Because I’ve never used them. But I’ll research them a bit tonight.

kbreit replied:
I haven't made this change yet, but I will.

felixfontein commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -127,6 +130,26 @@ def define_protocol(self):
+                    new = { snake_k: data[k] }
+                    items[snake_k] = self.sanitize_keys(data[k])
+            return items
+        elif isinstance(data, list):
```
Instead of `list`, use `Sequence`. Similarly, replace `dict` above by `Mapping`. (You'll need `from ansible.module_utils.common._collections_compat import Mapping, Sequence` to be able to do that.)

kbreit replied:
I haven't made this change yet, but I will.

felixfontein commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -127,6 +130,26 @@ def define_protocol(self):
+        if isinstance(data, dict):
+            items = {}
+            for k, v in data.items():
+                try:
```
Why not something like:
```.py
  new_key = self.key_map.get(k)
  if new_key is None:
    new_key = re.sub('([a-z0-9])([A-Z])', r'\1_\2', k).lower()
  items[new_key] = self.sanitize_keys(v)
```
(instead of the whole `try/except` block)

kbreit replied:
I haven't made this change yet, but I will.

felixfontein commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -127,6 +130,26 @@ def define_protocol(self):
+            items = {}
+            for k, v in data.items():
+                try:
+                    new = { self.key_map[k]: data[k] }
```
I don't think you need `new` anywhere.

felixfontein commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -127,6 +130,26 @@ def define_protocol(self):
+            for k, v in data.items():
+                try:
+                    new = { self.key_map[k]: data[k] }
+                    items[self.key_map[k]] = self.sanitize_keys(data[k])
```
`data[k]` is the same as `v`:
```suggestion
                    items[self.key_map[k]] = self.sanitize_keys(v)
```

felixfontein reviewed (commented):
I guessed yesterday that @kbreit == mrproper ;-) (or remembered that from somewhen earlier)

This is also a way to do it. What you could do is not specify a default for `output_version` (except in the description), and add code which prints out a deprecation warning in case it is not specified by the user that the current default is `old`, but it will change to `new` in Ansible 2.12. Then people can stick to the old names if they want (and get rid of the warning) if they explicitly set the value to `old` (or whatever it will be), and already now request the new format if they explicitly specify `new`.

kbreit commented:
@felixfontein My goal is to deprecate the camel case keys in Ansible 2.12 and get rid of the parameter altogether (or at least have it do nothing to avoid breaking playbooks, TBD). I’ve been trying to think of a friendly and graceful way to make this change. Here’s my current plan, but looking for feedback.

Ansible 2.9 - Introduce `output_format` and remove the `default`. However, it does stay with the snake case format if they don’t specify. I agree `old` and `new` isn’t optimal. Maybe `snake` and `camel`?
Ansible 2.10 - Introduce a warning if they don’t specify their format. No behavior change.
Ansible 2.11 - I’m thinking of changing the implicit default format to snake case in this version. But it does require playbook updates in two versions. First, to either override `output_format` or change playbook keys. Second, to remove `output_format` from playbooks in 2.12.
Ansible 2.12 - Remove camel case altogether. I may keep the `output_format` parameter and include a warning. Or just remove it and add a warning in 2.11.

felixfontein commented:
I agree that this is kind of annoying. Hmm. It's not exactly an easy problem :) Maybe just returning both keys for 2.8 - 2.11 and dropping the camel case for 2.12 is the best solution, even though until them the output is somewhat cluttered? Or just keep camel case?

@dagwieers do you have any ideas/comments/...?

kbreit commented:
@felixfontein I'm also trying to think of the proper way to merge the data structures. It's probably relatively straight forward if it's just a dictionary. I need to experiment to see if I need to do a recursive function to merge or if a single`update()` would work. But if there's a list, which there is most of the time, it's more complicated.

I can't guarantee the top level key is going to be different. It could be, lets say `network` so it wouldn't have any conversion. If the top level structure is a list, which it often is, I can't merge them so the list instead of 5 items long is 10. That would break data structures relating to things like firewall rules.

I'm sure there's one or two other ways I could merge them into a single data structure. If you have better ideas, I'm all ears (or eyes. we're reading).

felixfontein commented:
Since different keys can show up at all levels (but do not have to), you'd have to do a recursive merge. But that shouldn't be that much harder than your current `sanitize_keys` function (in fact, you need that one as well, to be called in some situations).

If there's a list (whose key is both snake and camel case), it's not a problem, you'll keep the list, but call the merge function for all its entries (depending on what they are). If the entries are neither lists nor dicts, nothing needs to be changed. If they are lists or dicts, you can do the recursive merge for these entries.

brianberzins commented:
Chiming in with a couple thoughts:

As someone that's done a lot of development work in the past, the difference between camelCasing and snake_casing is something that doesn't bother me much. As someone that is trying to *teach* people (primarily sysadmins) configuration management: consistency is extremely valuable. 

With regards to the plan forward: I personally wouldn't be bothered by a single, relatively painful, breaking change between versions. When we upgrade, we'll run through everything and update it once, then be done with it. I heavily suspect I'm in the minority opinion here.

Otherwise, as it sounds like you're planning:
1. only camelCase
1. enable both, defaulting to camelCase
1. enable both, defaulting to snake_case
1. only snake_case

It seems to me like the primary advantage of the extra steps is to just give people more time to decide when to deal with the conversion. Allowing them features of the next couple of releases in the mean time. Feels like a fair bit of complexity in the role-out to allow people more time to update cleanly. I guess that my view is "it's awfully nice of you to give people the extra time, but I don't think it's strictly necessary"

kbreit commented:
My main objection to a single painful break (rip off band-aid method) is Ansible uses a 3-4 release deprecation cycle so depending on that policy, I may not be able to take that approach.

What I may do is have 2.9 switch to snake-case. There will be an optional env var which sets camel case. Come 2.12 (or whenever I can deprecate it), I'll remove that. One release before, throw a lot of warnings. The objection of using a normal parameter (ex. `syntax_version`) is a playbook author needs to go update their playbooks to add it anyways, so why not just update while you're at it.

kbreit committed:
Add support for ANSIBLE_MERAKI_FORMAT env var
- If env var is set to 'camelcase' it will output camelcase
- Otherwise, will default to snakecase
- Added note to documentation fragment
- As of now, all module documentation needs to be updated

kbreit committed:
Fix pep8 errors and remove output_version args

kbreit committed:
Restructure check in exit_json so it actually works

felixfontein commented:
> My main objection to a single painful break (rip off band-aid method) is Ansible uses a 3-4 release deprecation cycle so depending on that policy, I may not be able to take that approach.

This, and then having a cycle of at least one version where both works also makes transitioning to a new version easier. You don't have to upgrade all Ansible installations at the same time without suddenly not being able to run playbooks with some of the installations and not with the others.

Also, some people don't update to every new version, but skip a few before updating. So it makes sense to have at least a short deprecation cycle. If 4 releases is a good cycle is another topic, but that's what the core team decided for.

> What I may do is have 2.9 switch to snake-case. There will be an optional env var which sets camel case. Come 2.12 (or whenever I can deprecate it), I'll remove that. One release before, throw a lot of warnings. The objection of using a normal parameter (ex. `syntax_version`) is a playbook author needs to go update their playbooks to add it anyways, so why not just update while you're at it.

Env vars aren't better/worse than regular options. Also, you can set an env fallback for an option, for example: `docker_host=dict(type='str', default='...', fallback=(env_fallback, ['DOCKER_HOST']))`; this allows the user to set this option by setting the `DOCKER_HOST` env variable instead. So you should really provide a module option (with env fallback) instead of adding an env-only option.

kbreit committed:
Add changelog fragment

kbreit committed:
Change output_format to a parameter with env var fallback
- ANSIBLE_MERAKI_FORMAT is the valid env var
- Added documentation

kbreit commented:
@felixfontein Turns out 'from ansible.module_utils.common.dict_transformations import camel_dict_to_snake_dict' is a thing which does exactly what I need so I am testing it out instead of 'sanitize_keys()'. This would resolve I think all your suggestions.

kbreit committed:
Convert to camel_dict_to_snake_dict() which is from Ansible
- Fixed integration tests

kbreit committed:
Fix yaml lint error

kbreit committed:
exit_json camel_case conversion handles no data
- exit_json would fail if data wasn't provided
- Updated 3 integration tests for new naming convention

kbreit committed:
convert_camel_to_snake() handles lists and dicts
- The native Ansible method doesn't handle first level lists
- convert_camel_to_snake() acts simply as a wrapper for the method
- There maybe a situation where nested lists are a problem, must test
- Fixed integration tests in some modules

kbreit committed:
A few integration test fixes

kbreit committed:
Convert response documentation to snake case

kbreit committed:
Rebase

Qalthos commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -344,8 +371,16 @@ def exit_json(self, **kwargs):
+        else:
+            if 'data' in self.result:
+                try:
+                    # self.fail_json(msg="here", converted=self.convert_camel_to_snake(self.result['data']))
```
What is this doing here?

kbreit replied:
Meraki responds with camel case and based on Ansible's standards, I want to show keys as snake case. So in the `exit_json()` method, this is where I'm doing the conversion for display.

kbreit replied:
Removed comment

kbreit committed:
Remove comment

kbreit commented:
@Qalthos I updated this one to fix the conflict and remove the commented line. Please review. Thank you!
